src/passes/ifStatementTempVarPostpender.ts

```diff
@@ -0,0 +1,193 @@
+import {
+  Block,
+  FunctionCall,
+  Identifier,
+  IfStatement,
+  UncheckedBlock,
+  VariableDeclarationStatement,
+} from 'solc-typed-ast';
+import { AST } from '../ast/ast';
+import { ASTMapper } from '../ast/mapper';
+import { TranspileFailedError } from '../utils/errors';
+import { CairoFunctionDefinition } from '../ast/cairoNodes';
+import { createBlock, createCairoTempVar } from '../utils/nodeTemplates';
+import { hasPathWithoutReturn } from '../export';
+
+/*
+ * This pass does a live variable analysis of the code while simultaneously
+ * postpending tempvars of the live variables at the end of each if branch and
+ * after each IfStatement in order to prevent against revoked references.
+ *
+ * A variable is 'live' at a line if there is a reference to the variable after
+ * the line in the control flow for which there is no matching declaration.
+ *
+ * This following:
+ *
+ ********************
+ * // y is not live
+ * let y = 0
+ *   // y is live
+ *   if (x == 0) {
+ *   // y is live
+ * } else {
+ *   // y is not live
+ *   let y = 4;
+ *   // y is live
+ * }
+ * g(y) // reference to y
+ ********************
+ *
+ * Becomes:
+ *
+ ********************
+ * // y is not live
+ * let y = 0
+ * // y is live
+ * if (x == 0) {
+ *   // y is live
+ *   tempvar y = y;
+ * } else {
+ *   // y is not live
+ *   let y = 4;
+ *   // y is live
+ *   tempvar y = y;
+ * }
+ * tempvar y = y;
+ * g(y) // reference to y
+ ********************
+ */
+
+export class IfStatementTempVarPostpender extends ASTMapper {
+  private liveVars: Set<string> = new Set();
+
+  // Function to add passes that should have been run before this pass
+  addInitialPassPrerequisites(): void {
+    const passKeys: Set<string> = new Set<string>([
+      'An', // An creates cairoFunctionDefinition which this pass use
+    ]);
+    passKeys.forEach((key) => this.addPassPrerequisite(key));
+  }
+
+  visitCairoFunctionDefinition(node: CairoFunctionDefinition, ast: AST): void {
+    this.liveVars = new Set<string>();
+    [...node.implicits].forEach((implicit) => this.liveVars.add(implicit));
+    // analyse the block
+    if (node.vBody) {
+      this.dispatchVisit(node.vBody, ast);
+    }
+    // remove the parameters and implicits
+    [...node.implicits].forEach((implicit) => {
+      if (!this.liveVars.delete(implicit)) {
+        throw new TranspileFailedError(
+          `Implicit variable dead. ${implicit} in function ${node.name}.`,
+        );
+      }
+    });
+    this.liveVars = new Set<string>();
+  }
+
+  visitIdentifier(node: Identifier, _ast: AST): void {
+    if (
+      node.vReferencedDeclaration &&
```

Comment:
I think it's better to be specific, as in `node.vReferencedDeclaration !== undefined`